akka-actor/src/main/scala/akka/actor/dungeon/DeathWatch.scala

```diff
@@ -63,7 +64,19 @@ private[akka] trait DeathWatch { this: ActorCell =>
   protected def receivedTerminated(t: Terminated): Unit =
     terminatedQueued.get(t.actor).foreach { optionalMessage =>
       terminatedQueued -= t.actor // here we know that it is the SAME ref which was put in
-      receiveMessage(optionalMessage.getOrElse(t))
+      optionalMessage match {
+        case Some(custom) =>
+          // needed for stashing of custom watch messages to work (or stash will stash the Terminated message instead)
+          val originalMessage = currentMessage
```

Comment:
Hanging on to the original and restoring it after is not necessarily needed, I don't think any tests fails without it, and the field set to null pretty much immediately when this returns, added to avoid any surprise if changing the calling code. WDYT should we just skip it?

Reply:
we already have something similar to change the `currentMessage` in Timers.scala, see
```
              // this is important for stash interaction, as stash will look directly at currentMessage #24557
              actorCell.currentMessage = actorCell.currentMessage.copy(message = m)
```

and there we don't care about setting it back